### PR TITLE
Bug 1214761 - ADB pull fails for symlinks.

### PR DESCRIPTION
**Pull request by weideng**
When pulling symlinks, save the files they link to, the same as the c adb client does.

### CODE REVIEW

File: adb.js [hunk 1]
```diff
@@ -604,17 +571,25 @@ const ADB = {
...
+              fileData = new Uint8Array(tmpData);
               pkgData = pkgData.subarray(chunkSize, pkgData.length);
               currentPos += chunkSize;
-              remaining -= chunkSize;
               chunkSize = 0;
             }
             // handle partial chunk at the end of socket package
-            if (chunkSize > 0 && pkgData.length < chunkSize) {
-              fileData.set(pkgData, currentPos);
+            if (chunkSize > 0 && 0 < pkgData.length && pkgData.length < chunkSize) {
```
Comment (jryans) [inline adb.js]: Nit: Please flip this around to `pkgData.length > 0` so that the constant is on the right side

File: adb.js [hunk 2]
```diff
@@ -604,17 +571,25 @@ const ADB = {
             // handle full chunk
             if (chunkSize > 0 && pkgData.length >= chunkSize) {
               let chunkData = pkgData.subarray(0, chunkSize);
-              fileData.set(chunkData, currentPos);
+              let tmpData = new Uint8Array(currentPos + chunkSize);
+              if (fileData) {
+                tmpData.set(fileData, 0);
+              }
+              tmpData.set(chunkData, currentPos);
+              fileData = new Uint8Array(tmpData);
```
Comment (jryans) [inline adb.js]: I think this can just be `fileData = tmpData;`

File: adb.js [hunk 3]
```diff
@@ -604,17 +571,25 @@ const ADB = {
...
             // handle partial chunk at the end of socket package
-            if (chunkSize > 0 && pkgData.length < chunkSize) {
-              fileData.set(pkgData, currentPos);
+            if (chunkSize > 0 && 0 < pkgData.length && pkgData.length < chunkSize) {
+              let tmpData = new Uint8Array(currentPos + pkgData.length);
+              if (fileData) {
+                tmpData.set(fileData, 0);
+              }
+              tmpData.set(pkgData, currentPos);
+              fileData = new Uint8Array(tmpData);
```
Comment (jryans) [inline adb.js]: I think this can just be `fileData = tmpData;`

**Comment by weideng** [comment]
I will update the patch, thanks for your review.

**Comment by jryans** [comment]
Great, looks good to me! Thanks for working on this.

**Pull request merged**